### DISCUSSION TITLE
vweb: form support for `application/octet-stream` content type

### DISCUSSION THREAD

**Discussion by mcastorina** [vlang/v#11854] [Ideas]

**Comment by gamemaker1** [comment]
Hi there!

I'm interested in working on this, but it seems that VWeb already supports this. Here is what I tried (similar setup to #9750):

```v
import os
import vweb

pub struct App {
	vweb.Context
}

[post]
pub fn (mut app App) test() vweb.Result {
	os.write_file('test-output.zip', app.req.data) or { panic(err) }
	return app.text('File uploaded successfully')
}

fn main() {
	vweb.run<App>(App{}, 8000)
}
```

```bash
$ curl -H "Content-Type:application/octet-stream" --data-binary @test-input.zip http://localhost:8000/test
File uploaded successfully
```

I used a zipped up source code folder as `test-input.zip`.

The resulting `test-output.zip` file contained the correct files when unzipped.

It would be great if someone could let me know what exactly needs to be worked upon. 

Thanks!

**Comment by mcastorina** [comment]
@gamemaker1 thanks for picking this up and testing it out!

I think I was mistaken when opening this issue; I thought more had to be done to support `octet-stream`, however according to the [RFC](https://datatracker.ietf.org/doc/html/rfc2046#section-4.5.1)..

> The "octet-stream" subtype is used to indicate that a body contains arbitrary binary data.

So while vweb does not explicitly check for this header, it will read arbitrary binary data by default.

**Comment by gamemaker1** [comment]
Great! Does that mean this issue can be closed?

**Comment by JalonSolov** [comment]
Hmm... I think it might be working, but not as shown.  The `Content-Type` header is for when the server is sending something to the client, to tell it what type it is.  The client should use `Accept: application/octet-stream` to indicate the type of data it wants back from the server.

That said, I think the only thing left to check is whether or not `vweb` set the `Content-Type` header correctly in the response headers.

**Comment by gamemaker1** [comment]
I tried that out, and no, it does not set headers _**automatically**_ when responding:

```v
import os
import vweb

pub struct App {
	vweb.Context
}

[post]
['/upload']
pub fn (mut app App) upload_test() vweb.Result {
	os.write_file('uploaded-file.zip', app.req.data) or { panic(err) }
	return app.text('File uploaded successfully\n')
}

[get]
['/download']
pub fn (mut app App) download_test() vweb.Result {
	return app.ok(os.read_file('test.zip') or { panic('$err') })
}

fn main() {
	vweb.run<App>(App{}, 8000)
}
```

```
$ curl -iH "Accept:application/octet-stream" http://localhost:8000/download
HTTP/1.1 200 OK
Content-Type: text/plain
Content-Length: 218
Server: VWeb
Connection: close

... binary data ...
```

However, if we do `app.set_content_type('application/octet-stream')` right before calling `app.ok`, we get the proper response type:

```v
[get]
['/download']
pub fn (mut app App) download_test() vweb.Result {
	app.set_content_type('application/octet-stream')
	return app.ok(os.read_file('test.zip') or { panic('$err') })
}
```

```
$ curl -iH "Accept:application/octet-stream" http://localhost:8000/download
Content-Type: application/octet-stream
Content-Length: 218
Server: VWeb
Connection: close

...binary data...
```

Of course, in both these cases, the server does not check the `Accept` header when returning output.

A couple of suggestions:

- Could we add a method, say `app.file(file_path string)`, that allows you to send a file back and automatically infers the mime type?
- The server enforces compliance to the `Accept` header and checks the mime type of the payload before sending it along. This can be turned off by setting the `app.enforce_accept_header` variable to `false`. 

Here is a rough outline for the above:

```v
// #1 - vlib/vweb/vweb.v: line 169 (https://github.com/vlang/v/blob/master/vlib/vweb/vweb.v#L169)
// Respond to the HTTP request with a HTTP_OK status (200) and the file 
pub fn (mut ctx Context) file(f string) Result {
	file_data := os.read_file(f) or {}
	// TODO: mime library does not exist
	inferred_mime := mime.from_string(file_data) or {}
	ctx.send_response_to_client(inferred_mime, file_data)
	return Result{}	
}

// #2 - vlib/vweb/vweb.v: line 123 (https://github.com/vlang/v/blob/master/vlib/vweb/vweb.v#L123)
// Send a HTTP response to the client
[manualfree]
pub fn (mut ctx Context) send_response_to_client(mimetype string, res string) bool {
	if ctx.done {
		return false
	}
	ctx.done = true

	if ctx.enforce_accept_header {
		// This is only a rough outline; not tested
		// It is only supposed to work for one properly specified mime type, no regex or lists (for now)
		accept_mime_type := ctx.req.header.get('accept') or {}
		if mimetype != accept_mime_type {
			return false
		}
	}

	header := http.new_header_from_map({
		http.CommonHeader.content_type:   mimetype
		http.CommonHeader.content_length: res.len.str()
	}).join(ctx.header)

	mut resp := http.Response{
		header: header.join(vweb.headers_close)
		text: res
	}
	
	resp.set_version(.v1_1)
	resp.set_status(http.status_from_int(ctx.status.int()))
	
	send_string(mut ctx.conn, resp.bytestr()) or { return false }
	
	return true
}
```

All of the above will require a library that can figure out the mime type of a file (not from the file extension, from the **file data**). I couldn't find one from a quick search, please let me know if there is any such library.

I'd be happy to submit a PR to add the above changes, let me know what you think of them.

Thanks!

**Comment by JalonSolov** [comment]
I've written a a tool that figures out the type from the contents, ignoring the extension, twice now:  Once in Java, and once in Go.

Problem is that to do it properly you need several other things that don't exist for V, yet.  :-\  I do plan to write a V version once those other pieces exist, but that will be a while.

Until then, the best we have is https://github.com/nedpals/v-mime

**Comment by gamemaker1** [comment]
I guess we could use that: even [ExpressJS](https://github.com/expressjs/express/) (server/web app framework for nodejs) figures out the mime of the file using its extension.

**Comment by JalonSolov** [comment]
<nod>  Most things do.  I wrote the tool for a work project... we discovered that some people were using the "wrong" extension for various reasons, and management wanted to run a check to find out the actual contents.  It needed to be cross-platform, instead of relying on Unix `file` command or whatever.

**Comment by gamemaker1** [comment]
I would be happy to help out with implementing such a library, and other pieces it might require.

Also, what do you think of the changes to VWeb I suggested in [#9842 (comment)](https://github.com/vlang/v/issues/9842#issuecomment-895968749)?

**Comment by JalonSolov** [comment]
I think it's an excellent idea.  Here's official docs on the Accept header, though... multiple types can be specified.

https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Accept

I'm not sure what happen if the server doesn't support any of the types requested...

**Comment by gamemaker1** [comment]
> Multiple types can be specified.

Yes, it will need to be able to parse both comma-separated mime types as well as regex types (e.g.: `*/*`, `text/*`, etc.)

> I'm not sure what happen if the server doesn't support any of the types requested

Good point, I hadn't thought of that. Perhaps the `.file`, `.html`, `.json`, `.text` and `.ok` functions should return errors instead of results all the time, so the developer can forward the error when received:

```v
[get]
['/download']
pub fn (mut app App) download_zip_file() vweb.Result {
	return app.file('test.zip') or {
		// Assuming I get a `InvalidResponseContentType` error
		app.set_status(400, 'Bad request')
		app.text('Invalid "Accept" header set.')
	}
}
```

**Comment by JalonSolov** [comment]
Have to do more research.  It may not be an error, the server may just decide to send whatever it wants to send, whether it matches or not.  Then it would be up to the client to decide what to do with it.

**Comment by gamemaker1** [comment]
> It may not be an error, the server may just decide to send whatever it wants to send, whether it matches or not.

I think [RFC 7231 (pg 38)](https://datatracker.ietf.org/doc/html/rfc7231#page-38) backs that:

>   A request without any Accept header field implies that the user agent
>   will accept any media type in response.  If the **header field is
>   present in a request and none of the available representations for
>   the response have a media type that is listed as acceptable, the
>   origin server can either honor the header field by sending a 406 (Not
>   Acceptable) response or disregard the header field by treating the
>   response as if it is not subject to content negotiation.**

To enable developers to follow option 1 and send a `406 Not Acceptable` back, I think VWeb should return an error. The others can simply tell VWeb to ignore the `Accept` header by setting a config variable like `app.enforce_accept_header` to `false`.

What do you think?

**Comment by JalonSolov** [comment]
Sounds like a plan.  Could switch it and save one initialization... option defaults to off, and we keep current behavior.  If it is set to true, it returns 406 if it can't send any of the types in the Accept header.

**Comment by gamemaker1** [comment]
Sure, we could do that

**Comment by gamemaker1** [comment]
Hi!

I created a [library](https://github.com/gamemaker1/http-negotiator) for parsing `Accept` headers and checking if the 'accepted' media types match the ones supported by the server. I had a question though - would this need to be a part of `vlib` to use it in `vweb`, or are imports from vpm allowed? In case it needs to be part of `vlib`, how would I got about it?

Thanks!

**Comment by JalonSolov** [comment]
Regular imports from elsewhere should work fine - even things in `vlib` are imported.  There's been talk of moving `vweb` out of `vlib` too, though I don't expect that to happen... it's just too darn convenient to have it pre-installed with V.

As far as moving it into `vlib`, I'd say get it working first, show what it can do for V, and that decision can always come later.

**Comment by gamemaker1** [comment]
Sure, thanks. I'll try adding the accept-header check into `vweb` with the external library and let you know how it goes.

**Comment by mcastorina** [comment]
> Regular imports from elsewhere should work fine - even things in `vlib` are imported.

I thought `vlib` should not import modules from `vpm`?

> There's been talk of moving `vweb` out of `vlib`

Much of the `vweb` server handling code is being moved to `net.http`, though I don't believe `vweb` will be moved entirely. Having a builtin web framework is really convenient as @JalonSolov said.

**Comment by gamemaker1** [comment]
Hi! Update on adding the accept header check:

> I thought vlib should not import modules from vpm?

I added the `http_negotiator` module under the `x` namespace (instead of `gamemaker1`) in `vlib`, so it is imported as `x.http_negotiator`. Let me know if I should change this and put it somewhere else, like `net.http` or something.

> I'm not sure what happen if the server doesn't support any of the types requested

My proposed solution ([#9842 comment](https://github.com/vlang/v/issues/9842#issuecomment-896027190)) did not work for this problem:

Say we have a server that can only respond in `application/json`, and the client only accepts `text/html`. The library returns an error, saying `No preferred media types found`. Now, VWeb passes that error back to the developer. The developer receives that error, and attempts to return a `406 Not Acceptable` with `{ "error": "Server doesn't support the given mime type" }`. Only the dev cannot - because the `app.ok` method calls `ctx.send_response_to_client`, which checks the mime type again. It fails again, because `application/json` does not match `text/html`.

Two possible solutions (I think adding both will be benefitial):

1. We add a method (or maybe it already exists; I'm not sure) that allows you to send a response without a payload, as `406` errors don't seem to need a body.

2. We don't proactively check for the `Accept` header (anyways the default for `enforce_accept_header` is `false`): instead, we add one of the two functions (similar to how ExpressJS does it):

- `app.accepts(media_type string)`, which checks if the mime type given will be accepted by the client. Example usage:

```v
// Example with VWeb

module main

import vweb

pub struct App {
	vweb.Context
}

[get]
['/names']
pub fn (mut app App) get_names_of_languages() vweb.Result {
	if app.accepts('application/json') {
		return app.json('{"languages":["V", "Go", "Oberon", "Rust", "Swift", "Kotlin", "Python"]}')
	} else {
		app.set_status(406, 'Not Acceptable')
		// If solution 1 exists, we could do `return app.blank()` or something
		return app.ok('Invalid values in the `Accept` header\n')
	}
}

fn main() {
	vweb.run<App>(App{}, 8000)
}
```

- `app.get_accepted_media_types()`, which returns the most preferred media type. It can be used in a `match` statement:

```v
// Example with VWeb

module main

import vweb

pub struct App {
	vweb.Context
}

[get]
['/names']
pub fn (mut app App) get_names_of_languages() vweb.Result {
	match app.get_accepted_media_types() {
		'application/json' {
			return app.json('{"languages":["V", "Go", "Oberon", "Rust", "Swift", "Kotlin", "Python"]}')
		}
		'text/plain' {
			return app.text('V, Go, Oberon, Rust, Swift, Kotlin and Python')
		}
		else {
			app.set_status(406, 'Not Acceptable')
			// If solution 1 exists, we could do `return app.blank()` or something
			return app.ok('Invalid values in the `Accept` header\n')
		}
	}
}

fn main() {
	vweb.run<App>(App{}, 8000)
}
```

Please let me know if there is any better way of doing this. Thanks!

**Comment by JalonSolov** [comment]
https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/406
https://httpwg.org/specs/rfc7231.html#status.406

**Comment by gamemaker1** [comment]
> [developer.mozilla.org/en-US/docs/Web/HTTP/Status/406](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/406)
> [httpwg.org/specs/rfc7231.html#status.406](https://httpwg.org/specs/rfc7231.html#status.406)

1. Right, we could return a 406 directly. That's an acceptable minimal solution.

2. To be more developer-friendly, we could provide the following two functions (like ExpressJS does):
	a. add a `app.accepts(<media_type>)` function for the developer to check if a media type is accepted.
	b. add a `app.get_preferred_media_type()` function that allows for `match` handling by the developer.
	Code snippets for both are in my [message above](https://github.com/vlang/v/issues/9842#issuecomment-897354097).

Should we implement just 1 or 1 and 2?
